pkg/arvo/app/chat-cli.hoon

```diff
@@ -0,0 +1,1193 @@
+          %banish    (change-permission | +.job)
+        ::
+          %join      (join +.job)
+          %leave     (leave +.job)
+        ::
+          %bind      (bind-glyph +.job)
+          %unbind    (unbind-glyph +.job)
+          %what      (lookup-glyph +.job)
+        ::
+          %settings  show-settings
+          %set       (set-setting +.job)
+          %unset     (unset-setting +.job)
+          %width     (set-width +.job)
+          %timezone  (set-timezone +.job)
+        ::
+          %select    (select +.job)
+          %chats     chats
+          %help      help
+        ==
+    ::  +act: build action move
+    ::
+    ++  act
+      |=  [what=term app=term =out-action]
+      ^-  move
+      :*  ost.bowl
+          %poke
+          /cli-command/[what]
+          [our-self app]
+          out-action
+      ==
+    ::  +set-target: set audience, update prompt
+    ::
+    ++  set-target
+      |=  tars=(set target)
+      ^-  (quip move _this)
+      =.  audience  tars
+      [[prompt:sh-out ~] this]
+    ::  +create: new local mailbox
+    ::
+    ++  create
+      ::TODO  configurable security
+      |=  [security=chat-security =path gyf=(unit char)]
+      ^-  (quip move _this)
+      ::TODO  check if already exists
+      =/  =target  [our-self path]
+      =.  audience  [target ~ ~]
+      =^  moz  this
+        ?.  ?=(^ gyf)  [~ this]
+        (bind-glyph u.gyf target)
+      =-  [[- moz] this]
+      %^  act  %do-create  %chat-view
+      :-  %chat-view-action
+      :^  %create  path  security
+      ::  ensure we can read from/write to our own chats
+      ::
+      :-  ::  read
+          ?-  security
+            ?(%channel %journal)  ~
+            ?(%village %mailbox)  [our-self ~ ~]
+          ==
+      ::  write
+      ?-  security
+        ?(%channel %mailbox)  ~
+        ?(%village %journal)  [our-self ~ ~]
+      ==
+    ::  +change-permission: modify permissions on a local chat
+    ::
+    ++  change-permission
+      |=  [allow=? rw=?(%r %w %rw) =path ships=(set ship)]
+      ^-  (quip move _this)
+      :_  this
+      %+  murn
+        ^-  (list term)
+        ?-  rw
+          %r   [%read ~]
+          %w   [%write ~]
+          %rw  [%read %write ~]
+        ==
+      |=  =term
+      ^-  (unit move)
+      =.  path
+        =-  (snoc `^path`- term)
+        [%chat (target-to-path our-self path)]
+      =/  whitelist=(unit ?)
+        =-  ?~(- ~ `?=(%white kind.u))
+        ::TODO  +permission-of-target?
+        .^  (unit permission)
+            %gx
+            (scot %p our-self)
+            %permission-store
+            (scot %da now.bowl)
+            %permission
+            (snoc path %noun)
+        ==
+      ?~  whitelist
+        ~&  [%weird-no-permission path]
+        ~
+      %-  some
+      %^  act  %do-permission  %group-store
+      ^-  out-action
+      :-  %group-action
+      ?:  =(u.whitelist allow)
+        [%add ships path]
+      [%remove ships path]
```

Comment:
@loganallenc Implemented as discussed. Sanity check please? (:

Reply:
Will look into and test it starting now. Thanks for passing this to me.

Reply:
Why are you not supporting blacklists as well? Seems like you'd want to allow permissions to be changed for blacklists also.

Reply:
Read more closely, `=(u.whitelist allow)`. This just determines if we want to do an `%add` or a %remove` based on 1) if we're dealing with a whitelist (or a blacklist), and 2) if we want to allow or deny permission.

We're basically sniffing out the diagonal in this "truth table".

• | whitelist | blacklist
--- | --- | ---
**allow** | %add | %remove
**disallow** | %remove | %add

Reply:
Then what does this do? =-  ?~(- ~ `?=(%white kind.u))
I hate =- and consider it to be an anti-pattern. This seems to me to cause whitelist to be null when kind is not equal to %white.

Reply:
If the scry give an empty unit, we have no permissions for whatever reason. We're neither a blacklist nor a whitelist: we are unknown. In that case (`?~` checking for "empty unit"), just produce `~` for "empty unit".  
In the case where the scry _did_ give a permission in the unit value, we produce the unit of a flag that's true if the `kind` of the permission value (from `u`, as added to the subject by `=-`) is `%white`.

Sure, I should maybe be using `=;` (inverse `=/`) instead. I'll clarify this code a little bit with naming and a comment.

Reply:
thanks sir